DependencyInjection/Configuration.php

```diff
@@ -189,131 +191,215 @@ protected function configureClients(ArrayNodeDefinition $root)
-                    ->addDefaultsIfNotSet()
+        $children->arrayNode('cache')
+            ->canBeEnabled()
+            ->addDefaultsIfNotSet()
+                ->children()
+                    ->scalarNode('cache_pool')
+                        ->info('This must be a service id to a service implementing Psr\Cache\CacheItemPoolInterface')
+                        ->isRequired()
+                        ->cannotBeEmpty()
+                    ->end()
+                    ->scalarNode('stream_factory')
+                        ->info('This must be a service id to a service implementing Http\Message\StreamFactory')
+                        ->defaultValue('httplug.stream_factory')
+                        ->cannotBeEmpty()
+                    ->end()
+                    ->arrayNode('config')
+                        ->addDefaultsIfNotSet()
                         ->children()
-                            ->scalarNode('preserve_header')->defaultTrue()->end()
-                            ->scalarNode('use_default_for_multiple')->defaultTrue()->end()
+                            ->scalarNode('default_ttl')->defaultNull()->end()
+                            ->scalarNode('respect_cache_headers')->defaultTrue()->end()
                         ->end()
-                    ->end() // End redirect plugin
+                    ->end()
+                ->end()
+            ->end();
+        // End cache plugin
 
-                    ->arrayNode('retry')
-                    ->canBeDisabled()
-                    ->addDefaultsIfNotSet()
-                        ->children()
-                            ->scalarNode('retry')->defaultValue(1)->end()
-                        ->end()
-                    ->end() // End retry plugin
+        $children->arrayNode('cookie')
+            ->canBeEnabled()
+                ->children()
+                    ->scalarNode('cookie_jar')
+                        ->info('This must be a service id to a service implementing Http\Message\CookieJar')
+                        ->isRequired()
+                        ->cannotBeEmpty()
+                    ->end()
+                ->end()
+            ->end();
+        // End cookie plugin
 
-                    ->arrayNode('stopwatch')
-                    ->canBeDisabled()
-                    ->addDefaultsIfNotSet()
-                        ->children()
-                            ->scalarNode('stopwatch')
-                                ->info('This must be a service id to a service extending Symfony\Component\Stopwatch\Stopwatch')
-                                ->defaultValue('debug.stopwatch')
-                                ->cannotBeEmpty()
-                            ->end()
-                        ->end()
-                    ->end() // End stopwatch plugin
+        $decoder = $children->arrayNode('decoder');
+        if ($disableAll) {
+            $decoder->canBeEnabled();
+        } else {
+            $decoder->canBeDisabled();
+        }
```

Comment:
this construct is repeated quite a bit, but i can't think of a better solution. the global plugins need to be enabled by default, but on the client, we want to only see plugins that where explicitly configured.